src/rust/engine/fs/src/store.rs

```diff
@@ -429,31 +433,198 @@ mod local {
+  impl ByteStore {
+    pub fn new(address: String) -> ByteStore {
+      // TODO: Pass through a parallelism configuration option here.
+      ByteStore {
+        address: address,
+        env: Arc::new(grpcio::Environment::new(1)),
+      }
+    }
+
+    fn load_bytes_with<T: Send + 'static, F: Fn(&[u8]) -> T + Send + Sync + 'static>(
+      &self,
       fingerprint: Fingerprint,
-    ) -> Result<Option<Vec<u8>>, String> {
-      store
-        .load_file_bytes_with(fingerprint, Arc::new(|bytes: &[u8]| bytes.to_vec()))
-        .wait()
+      f: Arc<F>,
+    ) -> BoxFuture<Option<T>, String> {
+      let channel = grpcio::ChannelBuilder::new(self.env.clone()).connect(&self.address);
+      let client = bazel_protos::bytestream_grpc::ByteStreamClient::new(channel);
+      let stream = client.read(&{
+        let mut req = bazel_protos::bytestream::ReadRequest::new();
+        // TODO: Pass a size around, or resolve that we don't need to.
+        req.set_resource_name(format!("/blobs/{}/{}", fingerprint, -1));
+        req.set_read_offset(0);
+        // 0 means no limit.
+        req.set_read_limit(0);
+        req
+      });
+
+      // We shouldn't have to pass around the client here, it's a workaround for
+      // https://github.com/pingcap/grpc-rs/issues/123
+      future::ok(client)
+        .join(stream.map(|r| r.data).concat2())
+        .map(|(_client, bytes)| Some(bytes))
+        .or_else(|e| match e {
+          grpcio::Error::RpcFailure(grpcio::RpcStatus {
+                                      status: grpcio::RpcStatusCode::NotFound, ..
+                                    }) => Ok(None),
+          _ => Err(format!("Error making CAS read request: {:?}", e)),
+        })
+        .map(move |maybe_bytes| {
+          maybe_bytes.map(|bytes: Vec<u8>| f(&bytes))
+        })
+        .to_boxed()
+    }
+  }
+
+
+  impl super::ByteStore for ByteStore {
+    fn store_file_bytes(&self, _bytes: Vec<u8>) -> BoxFuture<Digest, String> {
+      unimplemented!()
     }
 
-    fn load_directory_proto_bytes(
-      store: &ByteStore,
+    fn load_file_bytes_with<T: Send + 'static, F: Fn(&[u8]) -> T + Send + Sync + 'static>(
+      &self,
       fingerprint: Fingerprint,
-    ) -> Result<Option<Vec<u8>>, String> {
-      store
-        .load_directory_proto_bytes_with(fingerprint, Arc::new(|bytes: &[u8]| bytes.to_vec()))
-        .wait()
+      f: Arc<F>,
+    ) -> BoxFuture<Option<T>, String> {
+      self.load_bytes_with(fingerprint, f)
+    }
+
+    fn record_directory(
+      &self,
+      _directory: &bazel_protos::remote_execution::Directory,
+    ) -> BoxFuture<Digest, String> {
+      unimplemented!()
+    }
+
+    fn load_directory_proto_bytes_with<
+      T: Send + 'static,
+      F: Fn(&[u8]) -> T + Send + Sync + 'static,
+    >(
+      &self,
+      fingerprint: Fingerprint,
+      f: Arc<F>,
+    ) -> BoxFuture<Option<T>, String> {
+      self.load_bytes_with(fingerprint, f)
+    }
+  }
+
+  #[derive(Clone)]
```

Comment:
Accidentally commented on the previous review, but: I believe in the case where we're cloning this, we could use the reference on the clone of the `Store` that we sent into the closure instead.

Reply:
Done